### ISSUE TITLE
feat: Phase 1 Python integration via a thin `mlxcel` client package over the OpenAI-compatible server

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#407]
## Problem / motivation

Add a pure-Python package (`mlxcel`) that lets users drive mlxcel from Python by building entirely on top of the **existing OpenAI-compatible HTTP server** (`mlxcel serve`). This is "Phase 1" of Python integration: zero changes to the Rust inference core, no native extension. It provides a vLLM/SGLang-style `LLM` entry point that either spawns and manages a local `mlxcel serve` process or connects to an already-running one, plus an escape hatch to the raw `openai` SDK for the full API surface.

Phase 2 (an in-process PyO3 native binding) is explicitly out of scope here and tracked separately. The package namespace is designed so Phase 2 can slot in later without breaking the public API.

Every comparable engine ships two Python surfaces, an in-process API and an OpenAI-compatible HTTP server:

- **vLLM**: `from vllm import LLM; llm.generate(...)` in-process, plus `vllm serve` for the OpenAI server.
- **SGLang**: `sgl.Engine(model_path=...)` in-process, plus `python -m sglang.launch_server` (FastAPI, OpenAI-compatible).
- **llama.cpp**: `llama-cpp-python` (`from llama_cpp import Llama`, a ctypes wrapper over `libllama`), plus `python -m llama_cpp.server` (OpenAI-compatible).

mlxcel already has the server half done (axum, OpenAI-compatible). So the fastest, lowest-risk Python path is a thin client over that server, which is exactly how vLLM and llama.cpp are used in production deployments. This issue delivers that client with first-class ergonomics so callers do not have to hand-roll subprocess management and HTTP plumbing.

## Proposed solution

**Scope: Server / API surface (mlxcel-server) plus Tooling / documentation.**

### Server contract this builds on (already implemented, verified)

No Rust changes are required. The client relies on these existing behaviors:

- **OpenAI-compatible routes** (`src/server/app.rs:94-186`): `/v1/chat/completions`, `/v1/completions`, `/v1/models`, plus llama-server `/completion`, `/tokenize`, `/detokenize`, and `/health`.
- **`serve` flags** (`ServeArgs` in `src/main.rs:747-1604`): `-m/--model <PATH_OR_REPO_ID>` (env `LLAMA_ARG_MODEL`), `--host` (default `127.0.0.1`, env `LLAMA_ARG_HOST`), `--port` (default `8080`, env `LLAMA_ARG_PORT`), `--api-key` (env `LLAMA_API_KEY`), `--api-key-file`, `--ctx-size`, `--n-predict`, `-a/--alias`, `--warmup`/`--no-warmup`.
- **Unix domain socket mode**: set `--port 0`, and `--host` is reinterpreted as the socket path (`src/main.rs:782-784`; bind logic in `serve_unix_socket` at `src/server/startup.rs:1039`, TCP in `serve_tcp` at `:1073`). There is no separate `--socket` flag. UDS and TCP are mutually exclusive based on the port value.
- **Startup ordering (important)**: the model is loaded (and warmed up, on by default) BEFORE the listener binds (`src/server/startup.rs`, provider constructed and warmed up, then `serve_unix_socket`/`serve_tcp` called last). Consequence: during a first run that downloads weights from Hugging Face, the socket/port may not accept connections for a long time, then flip to ready.
- **Readiness via `/health`** (`src/server/routes/health.rs:42-98`): returns HTTP **503** with `{"status":"loading model", ...}` while the model is not yet loaded, HTTP **503** `{"status":"no slot available"}` when all slots are busy, and HTTP **200** with `{"status":"ok","model":"<id>", ...}` only when ready to generate. This is the authoritative readiness signal.
- **Model id discovery** (`src/server/routes/models.rs:25-39`): `GET /v1/models` returns `{"object":"list","data":[{"id":"<alias-or-model-dir-name>","object":"model","created":<ts>,"owned_by":"user"}]}`. The `id` is the `--alias` if set, otherwise the model directory name. The client should auto-discover this id so callers never have to guess the `model` string.

### Goals

- An ergonomic, documented Python API to run generation against mlxcel with minimal boilerplate.
- Manage the `mlxcel serve` subprocess lifecycle correctly (spawn, wait-until-ready, stream logs, graceful shutdown, crash detection).
- Prefer a Unix domain socket on POSIX for low-overhead, port-conflict-free local IPC, with TCP fallback.
- Expose the raw `openai` client for the full API surface (tools, `response_format`, logprobs, multimodal) rather than reimplementing it.
- Be pip-installable from the repo and CI-tested without needing the Rust binary or a model.

### Proposed Python API

Single unified entry point, `mlxcel.LLM`, with two modes selected by arguments:

```python
import mlxcel

# Managed mode: spawn and manage a local `mlxcel serve` process.
with mlxcel.LLM("mlx-community/Qwen3-4B-4bit") as llm:
    text = llm.generate("def fib(n):", max_tokens=128, temperature=0.7)

    for delta in llm.stream("Write a haiku about autumn"):
        print(delta, end="", flush=True)

    reply = llm.chat([{"role": "user", "content": "Hello"}], max_tokens=64)

    print(llm.model)        # resolved model id (auto-discovered from /v1/models)
    print(llm.models())     # ["<id>"]
    ids = llm.tokenize("hello world")
    s = llm.detokenize(ids)

    # Escape hatch: full OpenAI surface (tools, response_format, logprobs, vision/audio).
    oai = llm.openai_client          # configured openai.OpenAI
    oai.chat.completions.create(model=llm.model, messages=[...], tools=[...])

# Connect mode: talk to an already-running server, no process management.
llm = mlxcel.LLM(base_url="http://localhost:8080/v1")     # TCP
llm = mlxcel.LLM(socket="/tmp/mlxcel.sock")               # Unix socket
```

Mode selection rules:
- If `base_url` or `socket` is provided, use **connect mode** (no subprocess).
- Otherwise `model` is required and the client uses **managed mode** (spawn `mlxcel serve`).
- Providing both a connect target and `model` is an error (ambiguous).

Async variant `mlxcel.AsyncLLM` mirrors the sync API using `openai.AsyncOpenAI` and `httpx.AsyncClient`. Sync is required for this issue; async is a secondary task within the same issue (can land as a fast follow if needed).

### Module layout

New top-level directory `python/` (the repo has no `python/`, `clients/`, or `bindings/` dir today; `examples/` is Rust):

```
python/
  pyproject.toml
  README.md
  src/mlxcel/
    __init__.py        # exports LLM, AsyncLLM, errors; package version
    _server.py         # ManagedServer: discovery, spawn, readiness, logging, shutdown
    _client.py         # LLM / AsyncLLM: OpenAI-backed wrappers + transport wiring
    _sampling.py       # kwargs -> OpenAI params mapping + extra_body passthrough
    errors.py          # MlxcelError, MlxcelServerError, MlxcelTimeoutError
    py.typed           # mark package as typed
  tests/
    test_client_mock.py    # httpx MockTransport, no real server
    test_lifecycle.py      # spawns a fake server script, exercises manager
    test_e2e.py            # @pytest.mark.e2e, needs MLXCEL_BIN + small model
    fake_server.py         # tiny stdlib HTTP server: /health 503->200, /v1/* canned
  examples/
    quickstart.py
    streaming.py
    structured_output.py
```

### Server lifecycle manager (`_server.py`, the core new logic)

`ManagedServer` responsibilities:

1. **Binary discovery**, in order: explicit `binary=` argument, then `MLXCEL_BIN` env var, then `shutil.which("mlxcel")`. If none found, raise `MlxcelServerError` with an actionable message (how to build/install the binary).
2. **Transport selection**: default to a Unix domain socket on POSIX (`--port 0`, `--host <socket_path>`), fall back to TCP (auto-pick a free port via an ephemeral bind) on Windows or when the caller passes `host=`/`port=` explicitly. Keep the socket path SHORT to avoid the `SUN_LEN` limit (about 104 bytes on macOS, 108 on Linux); prefer a short name under `$TMPDIR`/`/tmp`, and allow `socket=` override. (This length budget is a known sharp edge in this codebase.)
3. **Spawn** `mlxcel serve` with the resolved flags via `subprocess.Popen`, passing `--api-key` when configured, and redirecting stdout/stderr to pipes.
4. **Readiness wait** (authoritative): poll `GET /health` over the chosen transport until HTTP 200. Treat connection-refused / missing-socket and HTTP 503 as "still starting" and keep retrying with backoff. Abort early and raise `MlxcelServerError` (with the tail of captured stderr) if the child process exits before becoming ready. Use a generous, configurable `startup_timeout` (default on the order of minutes, because the first run may download weights). Watch the child process liveness throughout.
5. **Log forwarding**: stream child stderr lines (the server logs via `tracing`, including the "Starting mlxcel server on ..." line and download/load progress) to the Python `logging` logger `mlxcel.server` on a background thread, and retain the last N lines for error reporting. Note: that log line is emitted just before bind, so it is a useful progress hint but `/health` 200 remains the ready signal.
6. **Shutdown**: `terminate()` (SIGTERM), wait a grace period, then `kill()` (SIGKILL). Remove the socket file. Wire up `__enter__`/`__exit__`, `atexit`, and a best-effort `__del__` finalizer so a leaked handle still cleans up.
7. **Crash handling**: if the child dies mid-session, subsequent calls raise a clear `MlxcelServerError` carrying the stderr tail.

### Transport wiring (`_client.py`)

Use the OpenAI SDK with a custom `httpx` client so the same code path covers TCP and UDS:

```python
import httpx
from openai import OpenAI

# TCP
client = OpenAI(base_url=f"http://{host}:{port}/v1", api_key=api_key or "-")

# Unix domain socket
transport = httpx.HTTPTransport(uds=socket_path)
http_client = httpx.Client(transport=transport, base_url="http://mlxcel")
client = OpenAI(base_url="http://mlxcel/v1", api_key=api_key or "-", http_client=http_client)
```

Notes:
- Async uses `httpx.AsyncHTTPTransport(uds=...)` plus `AsyncOpenAI`.
- When injecting a custom `http_client`, the SDK's default timeouts/retries are replaced, so set sensible timeouts explicitly (or subclass `DefaultHttpxClient`).
- After readiness, call `/v1/models` once and cache `data[0].id` as `self.model`, so callers never pass a `model` string. Allow an explicit `model=` override.
- `tokenize`/`detokenize` call the native `/tokenize` and `/detokenize` routes (not part of the OpenAI schema) via the underlying `httpx` client.

### Sampling parameters and structured output

`_sampling.py` maps Python kwargs to OpenAI request fields: `max_tokens`, `temperature`, `top_p`, `stop`, `seed`, `presence_penalty`, `frequency_penalty`, `logit_bias`. Provide an `extra_body=` passthrough for mlxcel/llama-server-specific knobs (for example `top_k`, `min_p`, `repetition_penalty`, DRY settings) and expose `response_format=` for the server's llguidance-backed JSON-schema constrained decoding. Document which extras the server honors.

### Error model

- `MlxcelError` (base), `MlxcelServerError` (launch/health/crash failures, carries stderr tail), `MlxcelTimeoutError` (readiness timeout).
- HTTP/API errors propagate as native `openai` SDK exceptions for transparency (do not hide them); only lifecycle concerns get custom exceptions.

### Packaging and dependencies

- `python/pyproject.toml`, build backend `hatchling`. Pure-Python, wheel tag `py3-none-any`.
- Dependencies: `openai>=1.40`, `httpx>=0.27`. Python `>=3.9`.
- Distribution name `mlxcel`, import name `mlxcel`. PyPI publishing is a follow-up (note: confirm name availability at publish time).
- Ship `py.typed` (typed package). Configure `ruff` and `mypy`.

### Testing strategy

- **Unit tests** (CI-safe, no binary, no model): use `httpx.MockTransport` to fake server responses and assert `generate`/`stream`/`chat`/`tokenize` behavior, the kwargs-to-params mapping, model auto-discovery, and error mapping.
- **Lifecycle tests**: ship a tiny stdlib `fake_server.py` that binds the UDS/TCP listener and serves `/health` (503 then 200 after a short delay) plus canned `/v1/*` responses. Exercise `ManagedServer` end to end (discovery, spawn, health polling, ready, log capture, shutdown, and early-exit failure) without the Rust binary or a model.
- **E2E test**: `@pytest.mark.e2e`, skipped unless `MLXCEL_BIN` and a small model are available. Spawns the real `mlxcel serve`, generates, and asserts non-empty output. Documented in the package README.

### CI integration

Add `.github/workflows/python.yml` running on `ubuntu-latest`: `ruff` (lint/format), `mypy` (type check), and `pytest` (unit + lifecycle tests only; e2e skipped). This is fast and needs neither MLX nor the binary. It does not touch the existing Rust CI (which intentionally runs only cargo-deny and cargo-fmt; clippy/test run locally via `make verify`).

### Docs and examples integration

- Add `docs/python-client.md` (install, managed mode, connect mode, streaming, chat, structured output, the `openai_client` escape hatch, async usage, troubleshooting).
- Link it from `docs/README.md`, register it in the `mkdocs.yml` nav, and add a short "Python" section to the root `README.md`.
- Add runnable scripts under `python/examples/`.

### Non-goals (out of scope for Phase 1)

- No PyO3 / native in-process binding (Phase 2).
- No access to logits, hidden states, or KV cache (needs native or new endpoints).
- No multi-model hot-swap (the server hosts one model per process).
- No distributed/router orchestration driven from Python.
- No bundling the Rust binary inside the wheel (possible later follow-up).
- No embeddings wrapper (`/v1/embeddings` is not served today).

### Design decisions and open questions

- **Package name `mlxcel`** (vs a separate `mlxcel-client`): chosen so Phase 2's native backend can live under the same namespace (for example `pip install mlxcel[native]`, selected transparently by `LLM`) without an API break. Open to objection.
- **Default transport = Unix socket on POSIX**: lower overhead and no port conflicts, at the cost of the socket-path-length constraint. TCP remains available and is the Windows default.
- **Sync required, async secondary**: keeps the first deliverable focused; async mirrors the same wiring.
- **Optional future server ergonomic (NOT in this issue)**: a machine-readable readiness signal (for example printing the bound address as a JSON line on stdout once ready) would let the client skip `/health` polling. Deferred to keep Phase 1 free of core changes.

## Acceptance criteria

- [x] `pip install ./python` succeeds and `import mlxcel` works.
- [x] Managed mode: `with mlxcel.LLM("<small-model>") as llm: llm.generate(...)` returns non-empty text against a real spawned server (verified via the e2e test or manually, since the binary is not in CI).
- [x] Connect mode works against a separately started `mlxcel serve` over both TCP and a Unix socket.
- [x] Readiness handling is correct: a first-run model download does not cause a spurious failure, and a server that exits during startup raises `MlxcelServerError` with stderr context.
- [x] New `python.yml` CI job is green (unit + lifecycle tests pass).
- [x] Docs page exists and is linked from `README.md`, `docs/README.md`, and the mkdocs nav, with runnable examples under `python/examples/`.
- [x] No changes to the Rust inference core are required to satisfy the above.

### Task checklist

- [x] Scaffold `python/` package (pyproject, src layout, `py.typed`, ruff/mypy config).
- [x] `_client.py`: OpenAI-backed `LLM` with UDS/TCP httpx transport, model auto-discovery, sampling mapping, `tokenize`/`detokenize`, `.openai_client` escape hatch.
- [x] `_server.py`: `ManagedServer` (binary discovery, transport selection with short-socket-path handling, spawn, `/health` readiness polling with crash detection, stderr log forwarding, graceful shutdown).
- [x] Unify managed/connect modes in `LLM`, add context-manager + atexit cleanup and the error types.
- [x] Tests: unit (mock transport), lifecycle (fake server), e2e (marked, optional).
- [x] `.github/workflows/python.yml` (ruff + mypy + pytest unit/lifecycle).
- [x] Docs: `docs/python-client.md`, links in `README.md` / `docs/README.md` / `mkdocs.yml`, plus `python/examples/`.
- [x] (Secondary) `AsyncLLM` async client.

### Relation to Phase 2

Phase 2 will add an in-process PyO3 native binding (`mlxcel-py`, maturin, abi3, platform/accelerator-specific wheels) modeled on `mlx-lm`. This issue's `LLM` API and `mlxcel` namespace are intended to remain stable so the native backend can be introduced as an optional acceleration path later.

## Pre-submission checks

- [x] I searched existing issues and discussions and did not find a duplicate
